src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1303,10 +1310,13 @@ void RaftPart::processAskForVoteRequest(const cpp2::AskForVoteRequest& req,
     return;
   }
 
+  auto oldRole = role_;
   auto oldTerm = term_;
   // req.get_term() >= term_, we won't update term in prevote
   if (!req.get_is_pre_vote()) {
     term_ = req.get_term();
+    role_ = Role::FOLLOWER;
```

Comment:
This is also problermatic. Suppose `term_ == req.get_term()` and they are two competing candidates, you cannot simply convert the voter into follower even though its log is older.

Reply:
You mean only convert to follower when req.term is greater, but when terms are equal check normally?

Reply:
Right, you cannot convert a candidate into follower when seeing a RequestVote rpc of same term(but you should conver to 
 follower when seeing AppendEntries rpc of the same term).

Reply:
I don't think there is a big difference between AskForVote and AppendLog when term are same. Think of this case:
ABCDE are on same term and all of them start election, let's say A start first and convert BCDE into follower, there could be two result:
1. A is elected, BC voted for A, but DE rejected because their logs are more up-to-date. But this is ok, because logs on   DE is only minority.
2. A is not elected, in this case, one of BCDE will start a new round election.

Reply:
when term equal, could become leader only when this peer vote for the candidate.

Reply:
Yeah, if leader of same term step down, that would be quite unstable. Fixed

Reply:
Suppose in a cluster of 5 noedes: A B C D E;
A, B both starts an election at the same term 5, B's log if newer than A,
C, D, E voted for A, but before A converting to leader, B send a RequestVote rpc to A, and if we convert A to a follower,
B will also failed the election at term 5 because C, D, E had voted for A at term 5,
we will have to start a new election. 

We must process differently on seeing same term for RequestVote and AppendEntries rpc request. The reason is that, when you see an AppendEntries rpc of the same term, that mean a leader has benn elected, so you must step follower if your are not; but seeing same term int RequestVote rpc request does not means that a leader has been elected for this term, so, don't do it.

Reply:
Another very very very important problem: **reset `votedAddr_` when you update `term_`**.  

It's disastrous to just update `term_` but keep the value of `votedAddr_` from older term, a cluster might takes very long time to vote a leader.

Think we must do a thoughly check for this in our Raft implementation.

Reply:
> Another very very very important problem: reset votedAddr_ when you update term_

Why? We have discussed this issue in pre-vote, see the line 1351. 

`voteAddr_` and `votedTerm_` should check together, and they are only updated when voted for others.

Reply:
> 

Oh, I forgot `votedTerm_`